spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -400,34 +430,65 @@ private void doIfTransaction(Context ctx, Consumer<ReactiveFirestoreResourceHold
 	}
 
 	private <T> Write createUpdateWrite(T entity) {
-		String documentResourceName = buildResourceName(entity);
-		Document document = getClassMapper().entityToDocument(entity, documentResourceName);
-		return Write.newBuilder()
-				.setUpdate(document)
-				.build();
+		ResourceName resourceName = buildResourceName(entity);
+		Document document = getClassMapper().entityToDocument(entity, resourceName.getName());
+		Builder builder = Write.newBuilder().setUpdate(document);
+		if (resourceName.isGenerated()) {
+			builder.setCurrentDocument(Precondition.newBuilder().setExists(false).build());
+		}
+		return builder.build();
 	}
 
-	private <T> String buildResourceName(T entity) {
+	private <T> ResourceName buildResourceName(T entity) {
 		FirestorePersistentEntity<?> persistentEntity =
 				this.mappingContext.getPersistentEntity(entity.getClass());
 		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
 		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
-
-		return buildResourceName(persistentEntity, idVal.toString());
+		boolean generated = false;
+		if (idVal == null) {
+			generated = true;
+			if (idProperty.getType() != String.class) {
+				throw new FirestoreDataException("Automatic ID generation only supported for String type");
+			}
+			idVal = Internal.autoId();
+			persistentEntity.getPropertyAccessor(entity).setProperty(idProperty, idVal);
+		}
+		return new ResourceName(buildResourceName(persistentEntity, idVal.toString()), generated);
 	}
 
 	private String buildResourceName(FirestorePersistentEntity<?> persistentEntity, String s) {
 		return this.parent + "/" + persistentEntity.collectionName() + "/" + s;
 	}
 
-	private String getIdValue(Object entity, FirestorePersistentEntity persistentEntity) {
+	private Object getIdValue(Object entity) {
+		FirestorePersistentEntity<?> persistentEntity =
+						this.mappingContext.getPersistentEntity(entity.getClass());
 		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
 		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
 
-		return idVal.toString();
+		return idVal;
 	}
 
 	public FirestoreClassMapper getClassMapper() {
 		return this.classMapper;
 	}
+
+	private static class ResourceName {
```

Comment:
I see; I would avoid introducing a new class here if possible -

I think if you include it, you will have to write a lot more docs to describe what it means for future readers of this code - i.e. it would at least require a class-level javadoc, and also - `generated` is too general (i.e. generated what?); you may have to rename to `needsGeneratedId` or add comment; similarly `name` is too general as well - i.e. (name of what?)

It's not worth the burden of adding this class in my opinion if it is avoidable.